subsys/net/ip/tcp.c

```diff
@@ -2639,124 +2666,203 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
-				 * - Out of order data len > 0, new_len > 0, ignore the data in
-				 *   between
+			if (tcp_compute_new_length(conn, th, len) > 0) {
+				/* We do not implement half closed sockets, therefore
+				 * cannot accept new data in after sending our FIN, as
+				 * we are in sequence can send a reset now.
 				 */
-				if (new_len > 0) {
-					conn_ack(conn, + new_len);
-				}
-			}
+				net_stats_update_tcp_seg_drop(conn->iface);
 
-			/*
-			 * Only if our FIN flag has been acknowledged and all the data has been
-			 * received, it can go to the TIME_WAIT state
-			 */
-			if (FL(&fl, ==, (FIN | ACK),
-			    (th_ack(th) == conn->seq) && (th_seq(th) == conn->ack))) {
-				tcp_send_timer_cancel(conn);
-				conn_ack(conn, + 1);
-				tcp_out(conn, ACK);
-				next = TCP_TIME_WAIT;
-				NET_DBG("FIN acknowledged, going to TIME_WAIT_STATE");
-				verdict = NET_OK;
-			} else if (th &&
-				   (FL(&fl, ==, FIN, th_seq(th) == conn->ack) ||
-				    FL(&fl, ==, (FIN | ACK), th_seq(th) == conn->ack))) {
-				tcp_send_timer_cancel(conn);
-				conn_ack(conn, + 1);
-				NET_DBG("FIN not yet acknowleged, going to CLOSING state");
-				if (th_ack(th) != conn->seq) {
-					/* Not acknowledged the FIN flag yet, so resend */
-					tcp_out_ext(conn, (FIN | ACK), NULL, conn->seq - 1);
-				} else {
-					tcp_out(conn, ACK);
-				}
-				next = TCP_CLOSING;
-				verdict = NET_OK;
-			} else if (th && FL(&fl, ==, ACK, th_ack(th) == conn->seq)) {
+				next = tcp_enter_time_wait(conn);
+
+				tcp_out(conn, RST);
+				break;
+			}
+			if (FL(&fl, &, ACK, th_ack(th) == conn->seq)) {
+				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 "
+					"state seq %u, ack %u"
+					, conn->seq, conn->ack);
 				tcp_send_timer_cancel(conn);
 				next = TCP_FIN_WAIT_2;
 				verdict = NET_OK;
-				NET_DBG("FIN acknowledged, going to FIN_WAIT_2 state seq %u, ack %u"
-					, conn->seq, conn->ack);
+			}
+			/*
+			 * There can also be data in the message, so compute with the length
+			 * of the packet to check the sequence number of the FIN flag with tha ack
+			 */
+			if (FL(&fl, &, FIN, (th_seq(th) + len) == conn->ack)) {
+				conn_ack(conn, + 1);
+
+				/* State path is dependent on if the acknowledge is in */
+				if (next == TCP_FIN_WAIT_2) {
+					/* Already acknowledged, we can go further */
+					NET_DBG("FIN received, going to TIME WAIT");
+
+					next = tcp_enter_time_wait(conn);
 
-				/* Acknowledge any received data */
-				if (len > 0) {
 					tcp_out(conn, ACK);
-				}
-			} else if (th) {
-				NET_DBG("Staying in FIN_WAIT_1 flags 0x%x state seq %u, ack %u",
-					fl, conn->seq, conn->ack);
-				if (len > 0) {
+				} else {
+					/* Already acknowledged, we can go further */
+					NET_DBG("FIN received, going to CLOSING as no ACK has been "
+						"received");
 					tcp_send_timer_cancel(conn);
-					/* Send out a duplicate ACK, with the pending FIN flag */
-					tcp_out(conn, FIN | ACK);
+					tcp_out_ext(conn, FIN | ACK, NULL, conn->seq - 1);
```

Comment:
@rlubos: Do we need a send timer cancel here to avoid a buffer leak?

Reply:
isn't this done already just one line above? Or perhaps your comment is misplaced and should be about line 2734?